### PR TITLE
ICRC-103: list approval information

### PR DESCRIPTION
**Pull request by bogwar**
This is a draft for a standard to list existing allowances

### CODE REVIEW

File: ICRCs/ICRC-103/ICRC-103.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+## 5. Example Using Symbolic Values
+
+Assume that the ledger stores the following allowances, listed in lexicographic order:
+
+- **A1** = `((p0, s0), (p1, s1), a1)`
+- **A2** = `((p0, s0), (p2, s2), a2)`
+- **A3** = `((p0, s1), (p3, s3), a3)`
+- **A4** = `((p1, s1), (p4, s4), a4)`
+- **A5** = `((p1, s2), (p5, s5), a5)`
```
Comment (sea-snake) [inline ICRCs/ICRC-103/ICRC-103.md:83]: It's not completely clear what the variables are here (`p0`, `s1`, `a1`), I can guess it's principal 0, subaccount 1 and allowance 1 (not sure why this is repeated in the value for this example).

I would clarify here allowances are defined in this example in the data structure:
```(account: (<principal>,<subaccount>), spender: (<principal>, <subaccount>), <allowance-id>)```

Right now someone might confuse `s4` as spender 4 instead of subaccount 4.

File: ICRCs/ICRC-103/ICRC-103.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    to_spender : Account;
+    allowance : Allowance;
+}
+
+type Account = record {
+    owner : principal;
+    subaccount : opt blob;
+};
+
+type Allowance = record {
```
Comment (sea-snake) [inline ICRCs/ICRC-103/ICRC-103.md:49]: I would suggest combining the record from `ListAllowanceResult` and `Allowance` along the lines of:

```candid
type Allowance = record {
  from_account : Account;
  to_spender : Account;
  allowance : nat;
  expires_at : opt nat64;
}

ListAllowanceResult = vec Allowance;
```

I would assume that the from and to are more related to the allowance than it's expiry 😅

**Review by sea-snake — COMMENTED** [review]
Looks good! Left a few comments, sorry for the late review 😅

File: ICRCs/ICRC-103/ICRC-103.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+
+## 2. Metadata
+
+A ledger that implements ICRC-191 MUST include `record {name = "ICRC-191"; url = "https://github.com/dfinity/ICRC-1/standards/ICRC-191"}` as part of the output of `icrc1_supported_standards`.
+
+The endpoint introduced in this standard operates in two ways.  In the public version any principal can obtain the outstanding allowances of any other principal. In the private version, the allowances returned by the endpoint must have been issued by the caller (i.e. the caller is the principal controlling the source account of an allowance.)
+Which version of the standard is implemented by a ledger is specified through metadata which can be retrieved using `icrc1_metadata`.
+
+A ledger that implements ICRC-191 MUST return metadata `icrc191:public_allowances` of type Text (optional). The possible values are "true" if the allowance data is public and "false" otherwise.
```
[the remark this answers is not in the document]
Comment (bogwar) [reply ICRCs/ICRC-103/ICRC-103.md:22]: Good point -- there is not good reason why we should make the value optional. I'll remove that

File: ICRCs/ICRC-103/ICRC-103.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+* If `from_account` is not provided, it is instantiated as `Account{caller_principal, first_subaccount}`.  
+* If `from_account.subaccount` is not provided, it is instantiated with `first_subaccount`.
+
+If the ledger implements the private version of the standard, then the endpoint returns the empty list if `from_account.owner ≠ caller_principal`.
+
+Otherwise, the endpoint returns a list of records of the form `(account_1, account_2, allowance)` in lexicographic order, with `account_1.owner = from_account.owner`.
+ * If `prev_spender` is provided the list starts with the allowance immediately succeeding `(from_account, prev_spender)`.
+ * If `prev_spender` is not provided the list starts with the first allowance from `from_account`.
+
+The list is limited to at most `take` entries or a maximum number of entries (which is an internal constant of the ledger).
```
[the remark this answers is not in the document]
Comment (bogwar) [reply ICRCs/ICRC-103/ICRC-103.md:73]: It's good to keep the interfaces as consistent with one another as possible, so I'll add this metadata here as well.

File: ICRCs/ICRC-103/ICRC-103.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sea-snake) [reply ICRCs/ICRC-103/ICRC-103.md:22]: > Good point -- there is not good reason why we should make the value optional. I'll remove that

On another note, in ICRC-7 we also had a query method to get the same value directly without having to fetch and parse the whole collection metadata.

File: ICRCs/ICRC-103/ICRC-103.md (hunk 2 above)
Comment (bogwar) [reply ICRCs/ICRC-103/ICRC-103.md:49]: The type Allowance was already defined in the ICRC-2 standard and here I'm reusing. I could introduce yet another type for the flatened entries in the response (as you suggest), but is there an advantage to doing this?

Comment (sea-snake) [reply ICRCs/ICRC-103/ICRC-103.md:49]: Ah that makes sense, I don't see any clear advantage for either. I suggested the flattened structure mostly to simplify the response data structure for consumers. 

My main concern would be a possible confusion due to duplicate nested `allowance` properties, where one property refers to the structure and the inner property refers to the actual amount.

File: ICRCs/ICRC-103/ICRC-103.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (bogwar) [reply ICRCs/ICRC-103/ICRC-103.md]: I've called it `icrc103_max_returned_allowances`. Do you have a better suggestion?

File: ICRCs/ICRC-103/ICRC-103.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (bogwar) [reply ICRCs/ICRC-103/ICRC-103.md]: Is this icrc7_collection_metadata?  Why "collection"? :)

File: ICRCs/ICRC-103/ICRC-103.md (hunk 1 above)
Comment (bogwar) [reply ICRCs/ICRC-103/ICRC-103.md:86]: I've added this clarification

File: ICRCs/ICRC-103/ICRC-103.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sea-snake) [reply ICRCs/ICRC-103/ICRC-103.md]: > I've called it `icrc103_max_returned_allowances`. Do you have a better suggestion?

Maybe something along the lines of `icrc103_allowances_max_take_value`, in ICRC-7 we use `icrc7_max_take_value` that covers all paginated methods but it probably makes more sense here to make it specific to the method (allowances).

Overall, it's a nitpick, I'm fine with either, let's make sure to clearly document which take value the method returns. In ICRC-7 we picked "max take value" over something like "max returned" in combination with documenting that any take value higher than the "max take value" results in the "max take value" being used.

File: ICRCs/ICRC-103/ICRC-103.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sea-snake) [reply ICRCs/ICRC-103/ICRC-103.md]: We decided at the time to clarify where each metadata belongs to by calling it either collection or token metadata but never metadata without collection or token. We could also have opted for metadata and token metadata, but we were concerned that this would make the documentation hard to follow.

File: ICRCs/ICRC-103/ICRC-103.md (hunk 1 above)
Comment (sea-snake) [reply ICRCs/ICRC-103/ICRC-103.md:86]: Clarification is missing one `(` or either has one `)` too many.

File: ICRCs/ICRC-103/ICRC-103.md (hunk 2 above)
Comment (bogwar) [reply ICRCs/ICRC-103/ICRC-103.md:54]: I see -- but then the same argument should apply to the Allowance type defined in ICRC-2? :)

File: ICRCs/ICRC-103/ICRC-103.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (bogwar) [reply ICRCs/ICRC-103/ICRC-103.md]: That's fine -- i like consistency, so I called it `icrc103_max_take_value`.

File: ICRCs/ICRC-103/ICRC-103.md (hunk 1 above)
Comment (bogwar) [reply ICRCs/ICRC-103/ICRC-103.md:103]: Good catch!

File: ICRCs/ICRC-103/ICRC-103.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (bogwar) [reply ICRCs/ICRC-103/ICRC-103.md]: sounds good -- I've added the method per your suggestion

File: ICRCs/ICRC-103/ICRC-103.md (hunk 2 above)
Comment (sea-snake) [reply ICRCs/ICRC-103/ICRC-103.md:54]: > I see -- but then the same argument should apply to the Allowance type defined in ICRC-2? :)

ICRC-2 doesn't have nested approvals, actually the did file doesn't even have the Approval type https://github.com/dfinity/ICRC-1/blob/main/standards/ICRC-2/ICRC-2.did

Comment (bogwar) [reply ICRCs/ICRC-103/ICRC-103.md]: Yes, that's definitely fair, so there's no really good reason to not flatten the entries in the list. I've implemented this in the latest version. Thanks a lot for keeping at it!

File: ICRCs/ICRC-103/ICRC-103.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+Which version of the standard is implemented by a ledger is specified through the `icrc103:public_allowances` metadata.
+
+A ledger that implements ICRC-103 MUST return the metadata entry `icrc103:public_allowances` of type `text`. The possible values are "true" if the allowance data is public and "false" otherwise.
+
+The number of allowances that the ledger can return in response to a query is limited by an implementation-specific maximum, specified through the `icrc103:max_take_value` metadata.
+
+A ledger that implements ICRC-103 MUST return the metadata entry `icrc103:max_take_value` of type `nat`, indicating the precise maximum number of allowances the ledger will return in response to a query.
+
+Metadata entries can be retrieved using method `icrc1_metadata` defined in the ICRC-1 standard as well as `icrc103_collection_metadata` defined below.
```
Comment (sea-snake) [inline ICRCs/ICRC-103/ICRC-103.md:28]: `icrc103_collection_metadata` is not needed, both ICRC-1 and ICRC-7 have a single metadata method (`icrc1_metada` and `icrc7_collection_metadata` respectively) that should also be used by extensions. In the past this topic of additional metadata methods came up in the WG, and the decisions was made that one method in the base standard suffices considering that entry keys are prefixed by their standard.

**Review by sea-snake — CHANGES_REQUESTED** [review]

File: ICRCs/ICRC-103/ICRC-103.md (hunk 5 above)
Comment (bogwar) [reply ICRCs/ICRC-103/ICRC-103.md:28]: I see -- so these are viewed as the base standards. I thought that `icrc7_collection_metadata` only returns metadata that starts with `icrc7:....`. You're saying that it would return metadata pertaining to any other extention of ICRC_7 that the ledger implements. 
Very good, I'll remove that.

**Comment by sea-snake** [comment]
One last comment, not sure what the correct method name should be.

On ICRC-37 we have `icrc37_get_token_approvals` and `icrc37_get_collection_approvals` (since we have approvals either per token or the whole collection). That would mean that the method name in ICRC-103 along those lines would be `icrc103_get_approvals`. 

Where there's two differences:
- "approvals" instead of "allowances", which makes sense since the ICRC-2 allowance method returns the allowance (nat) while this ICRC-103 returns (a list of) the whole approval that includes the allowance among other details like from/to and expiry.
- "get" instead of "list", this is a nitpick, no idea which one of the two would be most consistent across existing ICRC standards, probably requires checking existing ICRC standards in depth to pick one or the other.

**Comment by bogwar** [comment]
> One last comment, not sure what the correct method name should be.
> 
> On ICRC-37 we have `icrc37_get_token_approvals` and `icrc37_get_collection_approvals` (since we have approvals either per token or the whole collection). That would mean that the method name in ICRC-103 along those lines would be `icrc103_get_approvals`.
> 
> Where there's two differences:
> 
> * "approvals" instead of "allowances", which makes sense since the ICRC-2 allowance method returns the allowance (nat) while this ICRC-103 returns (a list of) the whole approval that includes the allowance among other details like from/to and expiry.
> * "get" instead of "list", this is a nitpick, no idea which one of the two would be most consistent across existing ICRC standards, probably requires checking existing ICRC standards in depth to pick one or the other.

For NFTs approvals is appropriate because there is no partial transfer so you list outstanding approvals. For fungible tokens allowance seems more suitable because you don't list the original approvals but rather whatever is left to be transferred because some partial `transfer_from` may have occurred... so that's more of an allowance. 
I don't have strong feelings about get or list so I'm happy to go with get for consistency. WDYT?

**Comment by sea-snake** [comment]
> For NFTs, approvals is appropriate because there is no partial transfer so you list outstanding approvals. For fungible tokens allowance seems more suitable because you don't list the original approvals but rather whatever is left to be transferred because some partial `transfer_from` may have occurred... so that's more of an allowance. I don't have strong feelings about get or list so I'm happy to go with get for consistency. WDYT?

Makes completely sense, also in ICRC-2 there's the allowance method while in ICRC-37 there's the is approved method indeed. So to keep it consistent along those lines, allowances is indeed the better option.

Just double checked most ICRC standards, can't find any method at first glance that use "list" in the method name, so let's go with "get".

**Review by sea-snake — APPROVED** [review]
